samtranslator/model/apigateway.py

```diff
@@ -85,3 +87,170 @@ def make_auto_deployable(self, stage, swagger=None):
         hash = generator.get_hash(length=40)  # Get the full hash
         self.Description = "RestApi deployment id: {}".format(hash)
         stage.update_deployment_ref(self.logical_id)
+
+
+class ApiGatewayAuthorizer(object):
+    _VALID_FUNCTION_PAYLOAD_TYPES = [None, 'TOKEN', 'REQUEST']
+
+    def __init__(self, api_logical_id=None, name=None, user_pool_arn=None, function_arn=None, identity=None,
+                 function_payload_type=None, function_invoke_role=None):
+        if function_payload_type not in ApiGatewayAuthorizer._VALID_FUNCTION_PAYLOAD_TYPES:
+            raise InvalidResourceException(api_logical_id, name + " Authorizer has invalid "
+                                           "'FunctionPayloadType': " + function_payload_type)
+
+        if function_payload_type == 'REQUEST' and self._is_missing_identity_source(identity):
+            raise InvalidResourceException(api_logical_id, name + " Authorizer must specify Identity with at least one "
+                                           "of Headers, QueryStrings, StageVariables, or Context.")
+
+        self.api_logical_id = api_logical_id
+        self.name = name
+        self.user_pool_arn = user_pool_arn
+        self.function_arn = function_arn
+        self.identity = identity
+        self.function_payload_type = function_payload_type
+        self.function_invoke_role = function_invoke_role
+
+    def _is_missing_identity_source(self, identity):
+        if not identity:
+            return True
+
+        headers = identity.get('Headers')
+        query_strings = identity.get('QueryStrings')
+        stage_variables = identity.get('StageVariables')
+        context = identity.get('Context')
+
+        if not headers and not query_strings and not stage_variables and not context:
+            return True
+
+        return False
+
+    def generate_swagger(self):
+        authorizer_type = self._get_type()
+        APIGATEWAY_AUTHORIZER_KEY = 'x-amazon-apigateway-authorizer'
```

Comment:
Nit: Instead of using this key all over in this map, why not assign the submap to a var
`apigw_auth_key = swagger[x-amazon-apigateway-authorizer]`